python/pyarrow/_dataset_parquet.pyx

```diff
@@ -741,7 +737,15 @@ cdef class ParquetFragmentScanOptions(FragmentScanOptions):
             thrift_string_size_limit=self.thrift_string_size_limit,
             thrift_container_size_limit=self.thrift_container_size_limit,
         )
-        return type(self)._reconstruct, (kwargs,)
+        return _reconstruct_parquet_fragment_scan_options, (kwargs,)
+
+
+def _reconstruct_parquet_fragment_scan_options(kwargs):
+    # __reduce__ doesn't allow passing named arguments directly to the
+    # reconstructor, hence this wrapper.
+    # In Cython >= 3.0.0, function binding is turned on by default, so
+    # a global static method is used (instead of a class method) for pickling.
```

Comment:
Would it work if you make `_reconstruct` above a `staticmethod` instead?

Reply:
Or it might also be possible to disable the `binding` option specifically for `_reconstruct`:
https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#locally

Reply:
`_reconstruct` as a staticmethod failed with both builtin pickle and cloudpickle, but I can't remember the exact error. I can grab it for you if you'd like. You're right that we should also be able to add a cython decorator to disable binding if that's preferred. I'll go ahead and update with this change.

Reply:
Which docs? Cython docs?

It works fine in pure Python, but it may be a Cython wart, in which case it would be nice to report an issue:
```python
>>> class C:
...:     @staticmethod
...:     def sm(): pass
...: 
>>> pickle.dumps(C.sm)
b'\x80\x04\x95\x15\x00\x00\x00\x00\x00\x00\x00\x8c\x08__main__\x94\x8c\x04C.sm\x94\x93\x94.'
>>> pickle.loads(pickle.dumps(C.sm))
<function __main__.C.sm()>
>>> pickle.loads(pickle.dumps(C().sm))
<function __main__.C.sm()>
```

Reply:
Nevermind.. it is possible to make this a staticmethod. I don't know why I thought it wasn't working prior. Will update.

Reply:
Ahh I remember what happened. This works for Cython 3, but not Cython 0.29.36.

With Cython 0.29.36, we just need to add the `binding=True` compiler option.

my_test.pyx
```
# cython : language_level = 3
import cython

cdef class Foo():
    @staticmethod
    @cython.binding(True)  # This works
    def _reconstruct():
        return Foo()

    def __reduce__(self):
        return Foo._reconstruct, tuple()


@staticmethod
def _reconstruct(x):
    return Bar(x)


cdef class Bar():
    @staticmethod  # This does not work
    def _reconstruct():
        return Bar()

    def __reduce__(self):
        return Bar._reconstruct, tuple()
```

Usage here:
```
In [1]: from my_test import Foo, Bar

In [2]: f = Foo()

In [3]: b = Bar()

In [4]: import pickle

In [5]: pickle.dumps(f)
Out[5]: b'\x80\x04\x95#\x00\x00\x00\x00\x00\x00\x00\x8c\x07my_test\x94\x8c\x10Foo._reconstruct\x94\x93\x94)R\x94.'

In [6]: pickle.dumps(b)
---------------------------------------------------------------------------
PicklingError                             Traceback (most recent call last)
Cell In[6], line 1
----> 1 pickle.dumps(b)

PicklingError: Can't pickle <built-in function _reconstruct>: it's not the same object as my_test._reconstruct
```